chalk-solve/src/clauses/builtin_traits/pointee.rs

```diff
@@ -0,0 +1,107 @@
+    self_ty: Ty<I>,
+    metadata: Ty<I>,
+) {
+    let interner = db.interner();
+    let trait_id = db.well_known_trait_id(WellKnownTrait::Pointee).unwrap();
+    let substitution = Substitution::from1(interner, self_ty);
+    builder.push_fact(TraitRef {
+        trait_id,
+        substitution: substitution.clone(),
+    });
+    let trait_datum = db.trait_datum(trait_id);
+    assert_eq!(
+        trait_datum.associated_ty_ids.len(),
+        1,
+        "Pointee trait should have exactly one associated type, found {:?}",
+        trait_datum.associated_ty_ids
+    );
+    let metadata_id = trait_datum.associated_ty_ids[0];
+    let alias = AliasTy::Projection(ProjectionTy {
+        associated_ty_id: metadata_id,
+        substitution,
+    });
+    builder.push_fact(Normalize {
+        alias,
+        ty: metadata,
+    });
+}
+
+fn metadata_for_ty<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    self_ty: Ty<I>,
+) -> Result<Option<Ty<I>>, Floundered> {
+    let interner = db.interner();
+    let unit = TyKind::Tuple(0, Substitution::empty(interner)).intern(interner);
+    let usize_ty = TyKind::Scalar(chalk_ir::Scalar::Uint(chalk_ir::UintTy::Usize)).intern(interner);
+    Ok(match self_ty.kind(interner) {
+        TyKind::Str | TyKind::Slice(_) => Some(usize_ty),
+        TyKind::Array(_, _)
+        | TyKind::Never
+        | TyKind::Closure(_, _)
+        | TyKind::FnDef(_, _)
+        | TyKind::Scalar(_)
+        | TyKind::Raw(_, _)
+        | TyKind::Function(_)
+        | TyKind::InferenceVar(_, TyVariableKind::Float)
+        | TyKind::InferenceVar(_, TyVariableKind::Integer)
+        | TyKind::Generator(_, _)
+        | TyKind::GeneratorWitness(_, _)
+        | TyKind::Ref(_, _, _) => Some(unit),
+        TyKind::Adt(id, subst) => {
+            let adt_datum = db.adt_datum(*id);
+            if adt_datum.kind != AdtKind::Struct {
+                return Ok(Some(unit));
+            }
+            let last_field_ty = adt_datum
+                .binders
+                .map_ref(|b| b.variants.clone())
+                .substitute(interner, subst)
+                .into_iter()
+                .take(1) // We have a struct so we're guaranteed one variant
+                .flat_map(|mut v| v.fields.pop())
+                .next()
+                .unwrap_or(usize_ty);
+            metadata_for_ty(db, last_field_ty)?
+        }
+        TyKind::Tuple(_, subst) => {
+            let last_field_ty = subst
+                .iter(interner)
+                .rev()
+                .next()
+                .map(|x| x.assert_ty_ref(interner).clone())
+                .unwrap_or(usize_ty);
+            metadata_for_ty(db, last_field_ty)?
+        }
+        TyKind::BoundVar(_)
+        | TyKind::AssociatedType(_, _)
+        | TyKind::OpaqueType(_, _)
+        | TyKind::Foreign(_)
+        | TyKind::Error
+        | TyKind::Placeholder(_)
+        | TyKind::Dyn(_)
+        | TyKind::Alias(_) => None,
```

Comment:
I guess instead of the current recursive code I should add a `forall<T> Normalize(<LastField as Pointee>::Metadata -> T) :- Normalize(<Adt as Pointee>::Metadata -> T)`, right?

Reply:
It didn't work. It returns `No possible solution` for
```
goal {
    Normalize(<(i32, str) as Pointee>::Metadata -> usize)
} yields {
    expect![["Unique"]]
}
```
The added clauses are:
```
for<> Implemented(Str: Pointee),
for<> Normalize(<Str as Pointee>::Metadata -> Uint(Usize)),
for<> Implemented(2<Int(I32), Str>: Pointee),
for<type> Normalize(<Str as Pointee>::Metadata -> ^0.0) :- Normalize(<2<Int(I32), Str> as Pointee>::Metadata -> ^0.0),
```
Which looks reasonable to me. Is chalk solver incapable of solving higher ranked clauses, or I missed something?

Reply:
I added them in the reverse direction 🤦 I also found out that it doesn't need qualifier and I can add `Normalize(<Adt as Pointee>::Metadata -> <LastField as Pointee>::Metadata) directly. It is now working.